checkmate/url/expand.py

```diff
@@ -0,0 +1,68 @@
+"""An implementation of Google Web Risk suffix/prefix expressions."""
+import re
+from urllib.parse import urlparse
+
+
+class ExpandURL:
+    """Expand a URL into variations for comparison.
+
+    As defined in: https://cloud.google.com/web-risk/docs/urls-hashing#suffixprefix_expressions
+    """
+
+    IPV4 = re.compile(r"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}")
+
+    @classmethod
+    def expand(cls, normalised_url):
+        """Expand a URL into many versions for comparison.
+
+        This will:
+
+         * Remove the scheme and "://"
+         * Progressively remove parts from the start of the domain
+         * Progressively simplify the path
+
+        :param normalised_url: A pre-normalised URL to vary
+        :returns: A generator of domain suffix, path prefix variations
+        """
+        parsed = urlparse(normalised_url)
+
+        path_variants = tuple(cls._vary_path(parsed.path, parsed.query))
+
+        for host in cls._vary_hostname(parsed.netloc):
+            for path in path_variants:
+                yield host + path
+
+    @classmethod
+    def expand_single(cls, normalised_url):
+        """Create a single expansion of the URL.
+
+        This allows you to create the least simplified version of the
+        expansions to use to create rules for comparing with.
+        """
+        parsed = urlparse(normalised_url)
+
+        return normalised_url.replace(parsed.scheme + "://", "")
+
+    @classmethod
+    def _vary_hostname(cls, hostname):
+        yield hostname
+
+        if cls.IPV4.match(hostname):
+            return
+
```

Comment:
These functions are awful, but they work. It would be _very_ good to go over these again at some point to create an algorithm that makes more sense.